.eslintrc.json

```diff
@@ -97,7 +97,34 @@
     "spaced-comment": [ "error", "always", { "exceptions": [ "*", "!" ] } ],
     "switch-colon-spacing": [ "error", { "after": true, "before": false } ],
     "unicode-bom": [ "error" ],
-    "valid-jsdoc": [ "error", { "requireParamDescription": false, "requireReturnDescription": false, "requireReturn": false, "prefer": { "returns": "return" } } ],
+    "valid-jsdoc": [ "error", {
+      "requireParamDescription": false,
+      "requireReturnDescription": false,
+      "requireReturn": false,
+      "prefer": {
+        "abstract": "virtual",
+        "augments": "extends",
+        "constructor": "class",
```

Comment:
I don't think these are the same?

It's been useful to differentiate in documentation between the class (as a whole) and the specific method signature of the constructor. We do this for PHP code with Doxygen, and currently for (the majority of) our code with JSDuck, in which `@class` and `@constructor` are different concepts. If this relates to JSDoc, then we should postpone this until later. I would also expect JSDoc to be able to make the same distinction, though perhaps through different means.

Example:

* <https://doc.wikimedia.org/mediawiki-core/master/js/#!/api/mw.Uri>
* <https://doc.wikimedia.org/mediawiki-core/master/js/#!/api/mw.Title>

Reply:
Maybe we should use `@constructs` instead?

http://usejsdoc.org/tags-class.html
http://usejsdoc.org/tags-constructs.html

Reply:
In the present, that would produce an error with JSDuck.

Our current subset of tags are documented at <https://www.mediawiki.org/wiki/Manual:Coding_conventions/JavaScript#Tags>, with some projects defining additional custom tags and/or using JSDoc instead.

Once <https://phabricator.wikimedia.org/T187672> is further along,  we can revisit the defaults, but not yet.

We could, however, make a jsdoc-specific mixin, similar to the one we just landed for QUnit.

You'd use it like this:

```
{
  "extends:" [
    "wikimedia",
    "wikimedia/jsdoc"
  ]
}
```

Or something like that.

Meanwhile, we can also then add a stricter set of rules (rather than just a safe default subset) for JSDuck, under `wikimedia/jsduck`.

Reply:
Sounds good!